src/Catalog/CatalogCommitUtilities.cs

```diff
@@ -0,0 +1,457 @@
+                            // to use.  Updating the cursor here for each commit can be very slow.
+                            newCommitTimeStamp = commitBatchTasks.CommitTimeStamp;
+
+                            DequeueBatchesWhileMatches(processingBatches, batch => batch.MinCommitTimeStamp == newCommitTimeStamp.Value);
+
+                            commitBatchTasksMap.Remove(newCommitTimeStamp.Value);
+
+                            if (!commitBatchTasksMap.Any())
+                            {
+                                if (maxCommitTimeStamp > newCommitTimeStamp)
+                                {
+                                    // Although all commits for the current page have been successfully processed, the
+                                    // current CatalogCommitBatchTasks.CommitTimeStamp value is not the maximum commit
+                                    // timestamp processed.
+                                    newCommitTimeStamp = maxCommitTimeStamp;
+                                }
+                            }
+                        }
+                        else // Canceled or Failed
+                        {
+                            hasAnyBatchFailed = true;
+
+                            exceptions.AddRange(
+                                commitBatchTasks.BatchTasks
+                                    .Select(batch => batch.Task)
+                                    .Where(task => (task.IsFaulted || task.IsCanceled) && task.Exception != null)
+                                    .Select(task => ExceptionUtilities.Unwrap(task.Exception)));
+                        }
+                    }
+
+                    if (newCommitTimeStamp.HasValue)
+                    {
+                        front.Value = newCommitTimeStamp.Value;
+
+                        await front.SaveAsync(cancellationToken);
+
+                        Trace.TraceInformation($"{typeName}.{nameof(FetchAsync)} {nameof(front)}.{nameof(front.Value)} saved since timestamp changed from previous: {{0}}", front);
+                    }
+
+                    if (hasAnyBatchFailed)
+                    {
+                        DequeueBatchesWhileMatches(processingBatches, batch => batch.Task.IsCompleted);
+                    }
+
+                    hasAnyBatchBeenProcessed = true;
+
+                    EnqueueBatchesIfNoFailures(
+                        client,
+                        context,
+                        commitBatchTasksMap,
+                        unprocessedBatches,
+                        processingBatches,
+                        lastBatch,
+                        maxConcurrentBatches,
+                        processCommitItemBatchAsync,
+                        cancellationToken);
+                }
+
+                if (hasAnyBatchFailed)
+                {
+                    var innerException = exceptions.Count == 1 ? exceptions.Single() : new AggregateException(exceptions);
```

Comment:
>innerException [](start = 24, length = 14)

Will Application Insights have record off ALL of these exceptions?

Reply:
No, but I addressed this with a subsequent [commit](https://github.com/NuGet/NuGet.Services.Metadata/pull/401/commits/141ba285117dd9b043592d26eebcb6a97faad60c).